src/Data/Array/Mutable/Linear.hs

```diff
@@ -221,6 +223,23 @@ instance Consumable (Array a) where
   consume :: Array a #-> ()
   consume (Array arr) = arr `Unlifted.lseq` ()
 
+instance Data.Functor Array where
+  fmap (f :: a #-> b) arr =
+    size arr & \(arr', Ur s) -> go 0 s arr'
+   where
+    -- When we're mapping  an array, we first insert `b`'s
+    -- inside an `Array b` by unsafeCoerce'ing, and then we
+    -- unsafeCoerce the result to an `Array b`.
+    go :: Int -> Int -> Array a #-> Array b
+    go i s arr
+      | i == s =
+          Unsafe.coerce arr
+      | otherwise =
+          unsafeRead arr i
+            & \(arr', Ur a) ->
+                unsafeWrite arr' i (Unsafe.coerce (f a))
+            & \arr'' -> go (i+1) s arr''
```

Comment:
Nitpicks all of which you can ignore if you want to:
- We could `go` from `s` down to `0` which is a bit cleaner
- Eventually we might want to do this in the unlifted array and then wrap the `map` on unlifted arrays to lifted ones

Reply:
> We could go from s down to 0 which is a bit cleaner

That is confusing to me. I can easily understand it conciously, but it's just too ingrained to me that I should iterate the array from beginning to the end. If I move the other way around, I need to start from `len-1` and end at `-1` in this case, which is ugly to me. So if it's okay I'd prefer to keep it like it is :).

> Eventually we might want to do this in the unlifted array and then wrap the map on unlifted arrays to lifted ones

I did that, thanks! I couldn't even use `unsafeCoerce` because of the kind mismatch, so I had to use `unsafeCoerce#`, which was scary. I think since we're casting `MutableArray# a` to `MutableArray# b`, it should be fine.